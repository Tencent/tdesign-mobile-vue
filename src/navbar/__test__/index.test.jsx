import { h } from 'vue';
import { mount } from '@vue/test-utils';
import { describe, it, expect } from 'vitest';
import NavBar from '../navbar';
import { AppIcon as TIconApp } from 'tdesign-icons-vue-next';

const iconFunc = () => h(TIconApp);
describe('navbar', () => {
  describe('props', () => {
    it('fixed', async () => {
      const navbar = mount(<NavBar title="标题" />);
      expect(navbar.element.style.position).toBe('fixed');

      await navbar.setProps({ fixed: false });
      expect(navbar.element.style.position).toBe('relative');
    });

    it('leftArrow', () => {
      const navbar = mount(<NavBar title="标题" leftArrow />);
      expect(navbar.find('.t-navbar__left-arrow').exists()).toBeTruthy();
    });

    it('title', () => {
      const navbar = mount(<NavBar title="标题" />);
      expect(navbar.text()).toContain('标题');
    });

    it('titleMaxLength', async () => {
      const navbar = mount(<NavBar title="标题超过了六个字符" titleMaxLength={6} />);
      expect(navbar.text()).toContain('标题超过了六...');

      await navbar.setProps({ titleMaxLength: -1 });
      expect(navbar.text()).toContain('标题超过了六个字符');
      await navbar.setProps({ titleMaxLength: 10, title: '测试标题' });
      expect(navbar.text()).toContain('测试标题');
    });

    it('visible-animation', () => {
      const navbar = mount(<NavBar title="标题" visible={false} />);
      const el = navbar.element;
      expect(el.className).toContain('t-navbar--hide-animation');
    });

    it('visible', () => {
      const navbar = mount(<NavBar title="标题" visible={false} animation={false} />);
      const el = navbar.element;
      expect(el.className).toContain('t-navbar--hide');
    });
  });

  describe('events', () => {
    it('left-click', async () => {
      const fn = vi.fn();
      const wrapper = mount(<NavBar title="标题" onLeftClick={fn} />, {
        // 插槽名称必须要和组件内部保持一致'left'
        slots: {
          left: iconFunc,
        },
      });
      expect(wrapper.element).toMatchSnapshot();
      expect(wrapper.findComponent(TIconApp).exists()).toBeTruthy();
      const left = wrapper.find('.t-navbar__left');
      await left.trigger('click');
      expect(fn).toHaveBeenCalled();
    });

    it('right-click', async () => {
      const fn = vi.fn();
      const wrapper = mount(<NavBar title="标题" onRightClick={fn} />, {
        // 插槽名称必须要和组件内部保持一致'right'
        slots: {
          right: iconFunc,
        },
      });
      expect(wrapper.element).toMatchSnapshot();
      expect(wrapper.findComponent(TIconApp).exists()).toBeTruthy();
      const right = wrapper.find('.t-navbar__right');
      await right.trigger('click');
      expect(fn).toHaveBeenCalled();
    });
  });

  describe('functions', () => {
    it('left', () => {
      const test = () => '左侧内容';
      const navbar = mount(<NavBar title="标题" left={test} />);
      expect(navbar.text()).toContain('左侧内容');
    });

    it('capsule', () => {
      const navbar = mount(<NavBar title="标题" capsule={iconFunc} />);
      expect(navbar.find('.t-icon-app').exists()).toBeTruthy();
    });

    it('right', () => {
      const test = () => '右侧内容';
      const navbar = mount(<NavBar title="标题" right={test} />);
      expect(navbar.text()).toContain('右侧内容');
    });

    it('title', () => {
      const test = () => '标题';
      const navbar = mount(<NavBar title={test} />);
      expect(navbar.text()).toContain('标题');
    });
  });

  describe('slots', () => {
    it('left', () => {
      const slot = <div className="left-content">左侧内容</div>;
      const navbar = mount(<NavBar title="标题" />, {
        slots: {
          left: slot,
        },
      });
      expect(navbar.find('.t-navbar__left').find('.left-content').exists()).toBeTruthy();
    });

    it('capsule', () => {
      const navbar = mount(<NavBar title="标题" />, {
        slots: {
          capsule: iconFunc,
        },
      });
      expect(navbar.find('.t-navbar__capsule').exists()).toBeTruthy();
    });

    it('right', () => {
      const slot = <div className="right-content">右侧内容</div>;
      const navbar = mount(<NavBar title="标题" />, {
        slots: {
          right: slot,
        },
      });
      expect(navbar.find('.t-navbar__right').find('.right-content').exists()).toBeTruthy();
    });

    it('title', () => {
      const slot = <div>标题</div>;
      const navbar = mount(<NavBar />, {
        slots: {
          title: slot,
        },
      });
      expect(navbar.text()).toContain('标题');
    });
  });
});
