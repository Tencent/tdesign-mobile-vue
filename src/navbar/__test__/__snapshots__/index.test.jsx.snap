// Vitest Snapshot v1

exports[`navbar > events > left-click 1`] = `
<div
  class="t-navbar t-navbar--fixed t-navbar--visible-animation"
  style="position: fixed;"
>
  <div
    class="t-navbar____placeholder"
  />
  <div
    class="t-navbar__content"
  >
    <div
      class="t-navbar__left"
    >
      <!---->
      
      <svg
        class="t-icon t-icon-app"
        fill="none"
        height="1em"
        viewBox="0 0 24 24"
        width="1em"
      >
        <path
          d="M2 2h9v9H2V2zm2 2v5h5V4H4zm13.5 0a2.5 2.5 0 100 5 2.5 2.5 0 000-5zM13 6.5a4.5 4.5 0 119 0 4.5 4.5 0 01-9 0zM2 13h9v9H2v-9zm2 2v5h5v-5H4zm9-2h9v9h-9v-9zm2 2v5h5v-5h-5z"
          fill="currentColor"
        />
      </svg>
      
      <!---->
    </div>
    <div
      class="t-navbar__center"
    >
      <span
        class="t-navbar__center-title"
      >
        标题
      </span>
    </div>
    <!---->
  </div>
</div>
`;

exports[`navbar > events > right-click 1`] = `
<div
  class="t-navbar t-navbar--fixed t-navbar--visible-animation"
  style="position: fixed;"
>
  <div
    class="t-navbar____placeholder"
  />
  <div
    class="t-navbar__content"
  >
    <div
      class="t-navbar__left"
    >
      <!---->
      <!---->
      <!---->
    </div>
    <div
      class="t-navbar__center"
    >
      <span
        class="t-navbar__center-title"
      >
        标题
      </span>
    </div>
    <div
      class="t-navbar__right"
    >
      
      <svg
        class="t-icon t-icon-app"
        fill="none"
        height="1em"
        viewBox="0 0 24 24"
        width="1em"
      >
        <path
          d="M2 2h9v9H2V2zm2 2v5h5V4H4zm13.5 0a2.5 2.5 0 100 5 2.5 2.5 0 000-5zM13 6.5a4.5 4.5 0 119 0 4.5 4.5 0 01-9 0zM2 13h9v9H2v-9zm2 2v5h5v-5H4zm9-2h9v9h-9v-9zm2 2v5h5v-5h-5z"
          fill="currentColor"
        />
      </svg>
      
    </div>
  </div>
</div>
`;
